Fixed challenges achievement always displaying

hairlessbear opened the pull request:
This causes the challenge achievement badge and text to only show up when the user has won at least one challenge.

hairlessbear committed:
Fixed challenges achievement always displaying

hairlessbear committed:
Fixed challenge achievement always displaying

Challenge achievement will now only display if the user has won 1 or more challenges.

Alys commented:
@lemoness For your approval. This PR deletes the section indicated by the arrow below when there are no won challenges. 
Note though for comparison the way that unearned achievements are greyed out higher up in the list.

![screen shot 2015-04-24 at 4 44 31 pm](https://cloud.githubusercontent.com/assets/1495809/7313934/56eb374a-eaa2-11e4-81ac-d734e0ac65de.png)

hairlessbear commented:
Thanks for the screenshot and explanation @Alys, I'll try to be
clearer next time. I also want to mention that the completed quests
achievement directly beneath the challenges achievement is completely
hidden if a user has completed zero quests. If you prefer the greyed out
style, should that one be changed too?

Alys commented:
Good point about the quest achievement being hidden too! That makes the challenges and quests achievements consistent with each other.

I personally don't have an opinion either way about grey versus invisible. :)

lemoness commented:
Hmm. It would probably make sense to be consistent one way or the other -
perhaps both the Challenges and the Quests should be greyed out, to
encourage users to investigate those features.

This would be in contrast with the "fun" badges, like the Savior of
Dilatory badge, that can't currently be earned by any specific action the
user takes. It makes sense for those to remain hidden. Getting a badge for
winning a challenge or completing a quest, though - those are both things
that the user can actively strive towards.

@sugarfiend Any thoughts?

On Fri, Apr 24, 2015 at 12:01 AM, Alys notifications@github.com wrote:

> Good point about the quest achievement being hidden too! That makes the
> challenges and quests achievements consistent with each other.
> 
> I personally don't have an opinion either way about grey versus invisible.
> :)
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5066#issuecomment-95825750.

hairlessbear committed:
Greyed-out challenges achievement

hairlessbear committed:
Added a missing dot

hairlessbear commented:
Turns out I was wrong about the quest achievement not showing up. It doesn't show up if you're viewing someone else's profile, but it shows up greyed-out when looking at your own profile. This behavior is identical to the behavior of the other greyed-out achievements, I just missed it the first time. 

Anyway, I changed the code so the challenges achievement now behaves like the other greyed-out achievements. @lemoness, what do you think?

![selection_004](https://cloud.githubusercontent.com/assets/7059890/7329775/4e4b2104-eab1-11e4-9bd1-196d477fcff6.png)

crookedneighbor commented:
Oh weird. I had forgotten that [] are both truthy and falsy. http://stackoverflow.com/questions/4226101/conflicting-boolean-values-of-an-empty-javascript-array#answer-4226263

Your length solution is :+1:

crookedneighbor commented on website/views/shared/profiles/achievements.jade:
```diff
@@ -146,8 +146,8 @@ div(ng-if=':: profile.achievements.originalUser')
   // This is a very strange icon to use. revisit
-  .achievement.achievement-karaoke(ng-show='profile.achievements.challenges')
-  div(ng-class='{muted: !profile.achievements.challenges}')
+  .achievement.achievement-karaoke(ng-show='profile.achievements.challenges.length')
```
While we're making changes anyway, can we make this `ng-show` an `ng-if` instead?

hairlessbear replied:
Sure, but most of the other achievements use `ng-show` on their achievement icons. If I'm going to change one, I should just go ahead and change them all, right?

Also, just to increase my understanding, why do you prefer `ng-if` over `ng-show` if they have the same result?

crookedneighbor replied:
Also, it should be safe to add the [one time binding marking](https://docs.angularjs.org/guide/expression#one-time-binding) (`::`) to both of them, so:

``` jade
.achievement.achievement-karaoke(ng-if='::profile.achievements.challenges.length')
div(ng-class='::{muted: !profile.achievements.challenges.length}')
```

crookedneighbor replied:
`ng-show` and `ng-hide` control the `display` property of the element with classes, so they're still present on the dom. `ng-if` removes it from the dom entirely if it doesn't satisfy the condition, which technically improves the performance of the web-app, though the performance boost would probably be so small, it'd be negligible. There are some tricky considerations to make when using `ng-if` over `ng-show`, but they don't apply to this circumstance. 

Here's a stackoverflow page that goes into some more depth: http://stackoverflow.com/questions/19177732/what-is-the-difference-between-ng-if-and-ng-show-ng-hide

hairlessbear replied:
Thanks for the info! Would you like me to make those changes (`ng-show` -> `ng-if` and one time bindings) on all of the relevant achievements?

crookedneighbor replied:
It's not required for this PR, but if you're feeling in the mood to do something tedious like that, I won't stop you :wink:

hairlessbear replied:
I'll add that to my list of "things to do when I should really be working on something else" :P

crookedneighbor replied:
I have so many of those! 

Ping me when the ng-if change is made for this achievement and I'll merge it in.

hairlessbear committed:
Changed ng-show to ng-if and added one-time bindings

hairlessbear commented:
ping @crookedneighbor, I changed the `ng-show` to `ng-if` and added one-time bindings in all relevant cases within `achievements.jade`. Unless I made some terrible mistake, it should be ready to go!

crookedneighbor commented on website/views/shared/profiles/achievements.jade:
```diff
@@ -164,21 +164,21 @@ div(ng-if='profile.achievements.quests || user._id == profile._id')
   .achievement.achievement-snowball
   h5=env.t('annoyingFriends')
   small
     =env.t('annoyingFriendsText', {snowballs: "{{profile.achievements.snowball}}"})
```
You can add the one time bindings to `snowballs: "{{::profile.achievements.snowball}}"` too.

crookedneighbor replied:
Same goes for spook dust and agricultural friends.

hairlessbear replied:
How bout this: are there any items of the form `profile.achievements.*` that I can't change? It seems like they can all be changed.

hairlessbear replied:
With the obvious exception of items like `{muted: !profile.achievements.quests}`, since the one time binding is already on `muted`.

crookedneighbor replied:
Probably? But I'm not absolutely sure, so I'll just merge it in as is now and look at the rest in more depth later. Not going to hurt anything to be slightly less optimized for a little while longer :smile:

crookedneighbor commented:
Noticed three more places that could use one time bindings. If you want to tackle those, feel free. Otherwise, I'll merge it in and do it myself. Just let me know either way.

hairlessbear commented:
I'm working on it right now. I also noticed a couple more `ng-show`s that could be turned into `ng-if`s. Should I create a new branch for those?

crookedneighbor commented:
@hairlessbear You can keep it on this branch and just submit a new PR if you want. I've merged this one in and notated it toward your next contributor tier.

hairlessbear commented:
Got it. Should have that up in a few minutes. I found a lot more than three places one-time bindings could be used, so hopefully I'm not doing something wrong.

crookedneighbor commented:
@hairlessbear The only time having a one time binding would be wrong is if that value could change after it gets loaded into the page for the first time. For instance, in the future, if there's a way to cast spook dust on a party member from any page, and someone cast it on themselves on the achievements page, their number count wouldn't go up. _I don't think there's anything like that now_, so it should be fine. 

Though, worst case scenario, we just remove that one time binding later, so it's not that big of a deal since it doesn't affect the site's usability in the meantime.

crookedneighbor commented:
^ let me be clear about the spook dust example. Their number count WOULD go up, it just wouldn't appear on the achievements page until they left the page and came back.

hairlessbear commented:
Cool, that's what I thought. But that means that every single instance of `user._id` could have a one-time binding, right? If so, I can go ahead and change that too.

The only ones I wasn't quite sure of were those contained within a `ng-repeat` (example: `ng-repeat='(k,v) in profile.achievements.quests'`). But from what you said, it sounds like those are fine to change as well.

hairlessbear commented:
I mean, technically, a quest could be completed while someone was looking at the achievement page, causing it to not show up, but that seems like fringe case at best.

crookedneighbor commented:
Yeah, that was the only case I could think of. I don't mind users having to either refresh or leave and come back for their quest info to update. If someone else feels strongly, they can speak up or remove the one-time binding.

hairlessbear commented:
Ok, I'm almost done. I hit a wrong button in vim and messed up some indentation, so I have to fix that first before I create the new PR.

lemoness commented:
@hairlessbear I like the look of the implementation! One question, though -
why are the greyed-out badges not displayed? Not saying they should be;
just curious.

On Sat, Apr 25, 2015 at 5:19 PM, hairlessbear notifications@github.com
wrote:

> Ok, I'm almost done. I hit a wrong button in vim and messed up some
> indentation, so I have to fix that first before I create the new PR.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5066#issuecomment-96298579.

hairlessbear commented:
@lemoness As far as I know, greyed-out badges have never been displayed. When a user hasn't earned badges, the text displays in greyscale, but the badges are hidden. All I did was make the challenges achievement conform to this style (it was always displayed as earned before); the rest are as they were.

By the way, if you do decide that you'd like greyed-out badges in addition to the greyed-out text, it looks like that would be [pretty easy](http://stackoverflow.com/questions/609273/convert-an-image-to-grayscale-in-html-css) to implement.

crookedneighbor commented:
@hairlessbear is correct. The idea is that you don't get the art for the badge until you actually earn the achievement.

lemoness commented:
Makes sense!

On Wed, Apr 29, 2015 at 2:49 PM, Blade Barringer notifications@github.com
wrote:

> @hairlessbear https://github.com/hairlessbear is correct. The idea is
> that you don't get the art for the badge until you actually earn the
> achievement.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/5066#issuecomment-97597749.

crookedneighbor commented:
^ I'm not in favor of gray scaling it. Too many weird browser issues to consider. :/
